Extend/Order_Controller.php

```diff
@@ -0,0 +1,137 @@
+<?php
+/**
+ * Shop System Plugins:
+ * - Terms of Use can be found under:
+ * https://github.com/wirecard/oxid-ee/blob/master/_TERMS_OF_USE
+ * - License can be found under:
+ * https://github.com/wirecard/oxid-ee/blob/master/LICENSE
+ */
+
+namespace Wirecard\Oxid\Extend;
+
+use \OxidEsales\Eshop\Core\Registry;
+use \Wirecard\Oxid\Extend\Payment_Gateway;
+use \Wirecard\Oxid\Extend\Order;
+use \Wirecard\PaymentSdk\Response\FailureResponse;
+use \Wirecard\PaymentSdk\Response\InteractionResponse;
+use \Wirecard\PaymentSdk\Entity\Redirect;
+
+/**
+ * Class Order
+ *
+ * @package Wirecard\Extend
+ *
+ * @mixin \OxidEsales\Eshop\Application\Model\Order
+ */
+class Order_Controller extends Order_Controller_parent
+{
+    /**
+     * Extends the parent init function and finalizes the order in case it
+     * was a Wirecard payment method
+     */
+    public function init()
+    {
+        parent::init();
+        $oConfig = Registry::getConfig();
+        $sWdPaymentRedirect = $oConfig->getRequestParameter('wdpayment');
+        $oSession = Registry::getSession();
+        $sWdSessionToken = $oSession->getVariable('wdtoken');
+
+        if (!empty($sWdSessionToken) && $sWdSessionToken == $sWdPaymentRedirect) {
+            $sShopBaseUrl = $oConfig->getShopUrl();
+            $sLanguageCode = Registry::getLang()->getBaseLanguage();
+
+            $aParams = array(
+                'lang' => $sLanguageCode,
+                'force_sid' => $oSession->getId(),
+                'stoken' => $oSession->getSessionChallengeToken(),
+                'actcontrol' => 'order',
+                'cl' => 'order',
+                'fnc' => 'execute',
+                'sDeliveryAddressMD5' => $this->getDeliveryAddressMD5(),
+                'challenge' => '',
+                'ord_agb' => '1',
+                'oxdownloadableproductsagreement' => '0',
+                'oxserviceproductsagreement' => '0',
+                'wdtoken' => $sWdSessionToken
+            );
+
+            $sParamStr = http_build_query($aParams);
+
+            $sNewUrl = $sShopBaseUrl . 'index.php?' . $sParamStr;
+            header('Location: '.$sNewUrl);
+        }
+    }
+
+    /**
+     * Redirects to the PaymentSDK if the payment is a Wirecard method
+     *
+     * @return string
+     *
+     */
+    public function execute()
+    {
+        $oConfig = Registry::getConfig();
+        $sWdPaymentRedirect = $oConfig->getRequestParameter('wdtoken');
+
+        $oBasket = $this->getSession()->getBasket();
+
+        $oOrder = oxNew(Order::class);
+        $sOrderId = Registry::getSession()->getVariable('sess_challenge');
+        $oOrder->load($sOrderId);
+
+        $sPaymentId = $oBasket->getPaymentId();
+
+        $sWdSessionToken = $this->getSession()->getVariable('wdtoken');
+
+        if (!$oOrder->isWirecardPaymentType($sPaymentId) ||
+         (!empty($sWdSessionToken) && $sWdSessionToken == $sWdPaymentRedirect)) {
```

Comment:
This second check could be put into separate method with the descriptive name. e.g. isWdPaymentMethodFinished